samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesValue.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.samza.operators.impl.store;
+
+/**
+ * An immutable pair of a value, and its corresponding timestamp.
+ *
+ * <p> Iterators on {@link TimeSeriesStore}s always return {@link TimeSeriesValue}s
+ *
+ * @param <V> the type of the value
+ */
+public class TimeSeriesValue<V> {
+  private final V value;
+  private final Long timestamp;
```

Comment:
Should this timestamp always match the timestamp in the key? If so, does it need to be in both the key and the value?

Reply:
There are no such requirements. 

- On the write side, you'd call `put(k, v, timestamp)`
- On the query side, you ask the store to give you all values for a key within a timestamp range. The return value is an iterator that returns the value along with the timestamp they were added to the store. 

There's no need to store the timestamp in both the key and the value.